### ISSUE TITLE
Running `magit-status' on a symlink to a regular file inside a git repo should open magit in the right directory

### ISSUE THREAD

**Issue by the-kenny** [magit/magit#2250]
Background: I use stow[1] for my dotfiles. That means I have a ```~/dotfiles``` directory which is managed via git and for example ```~/.zshrc``` is a symlink to ```~/dotfiles/zsh/.zshrc```.

If I open ```~/.zshrc``` in Emacs and run ```magit-status``` it used to open the status buffer for the ```~/dotfiles``` repository. Since ```2.1``` this isn't happening anymore. 

Searching lead to https://github.com/magit/magit/issues/703 which I believe is pre-2.1. Is the old behavior still available somehow? I dearly miss it.

**Comment by johnmastro** [comment]
Did you try `(setq vc-follow-symlinks t)`? (Or `(setq vc-follow-symlinks 'ask)` and answer yes when asked if you want to follow the link).

**Comment by the-kenny** [comment]
The default value on my setup was ```ask```, but it never asked. While writing this issue, I set it to ```t```. No difference in behavior.

**Comment by tarsius** [comment]
Magit uses `magit-toplevel` to determine the appropriate repository (more precisely the top-level directory of the working directory (except in a bare repository)). It does follow symlinks, but unlike `git rev-parse --show-toplevel` only when necessary, or when `file-visit-truename` is non-nil. When there is a symlink to a subdirectory to of a repository then that link has to be followed, a link to the top-level directory itself does not have to be followed.

However `magit-toplevel` is run inside `default-directory`, the `buffer-file-name` is irrelevant. `git rev-parse --show-toplevel` is also run in PWD, not "inside a regular file". Processes always run in a working directory, not inside a "working file".

So as far as `magit-toplevel` is concerned files don't matter, only `default-directory` does. So only links to directories matter, but not links to regular files.

Magit does not directly use `vc-follow-symlinks` but it is still relevant when using Magit, which is why I have added it to the `magit` custom group. In both VC as well as Magit this setting is honored at the time you visit a regular file.

`vc-follow-symlink` = `ask`/`t` only has an effect if `vc-find-file-hook` is a member of `find-file-hook`, so check that. Then, at the time you visit the file, you will be asked whether to follow the link into the repository (or it will be done automatically). You have to confirm this at this point. When you then call `magit-status` in the file-visiting buffer, then `default-directory` is inside the repository.

This also appears to work if the link itself is inside a repository (e.g. in your case if `~/` is a repository). It does however not work if the link is also tracked in that repository. Also note that VC caches whether a file is inside a repository and it might not always invalidate that cache properly.

----

TL;DR follow the symlink when you visit the regular file.
